File: dfs_bfs_backtracking/python/b10819.py
N = int(input())
nums = list(map(int, input().split()))
visited = [0] * N
max_num = 0

def bt(cnt, temp, latter):
    global N, max_num
    if cnt == N:
        if max_num < temp:
            max_num = temp
        return
    
    if cnt == 0:
        for i in range(N):
            visited[i] = 1
            bt(cnt + 1, temp, nums[i])
            visited[i] = 0
    else:
        for i in range(N):
            if not visited[i]:
                visited[i] = 1
                temp_sum = nums[i] - latter if nums[i] > latter else latter - nums[i]
                bt(cnt + 1, temp + temp_sum, nums[i])
                visited[i] = 0
    return

bt(0, 0, 0)
print(max_num)

File: dfs_bfs_backtracking/cpp/b10819.cpp
#include <iostream>
int num_list[8];
bool visited[8];
int N;
int max_value = 0;

void backtracking(const int& latter, const int& temp_max_value, const int& cnt) {
	if (cnt == N) {
		if (temp_max_value > max_value) {
			max_value = temp_max_value;
		}
		return;
	}

	if (cnt == 0) {
		for (int i = 0; i < N; ++i) {
			visited[i] = true;
			backtracking(num_list[i], 0, cnt + 1);
			visited[i] = false;
		}
	}
	else {
		for (int i = 0; i < N; ++i) {
			if (visited[i]) {
				continue;
			}
			visited[i] = true;
			int temp_sum = latter - num_list[i];
			temp_sum = temp_sum > 0 ? temp_sum : -temp_sum;
			backtracking(num_list[i], temp_max_value + temp_sum, cnt + 1);
			visited[i] = false;
		}
	}
	return;
}


int main() {
	std::cin >> N;
	for (int i = 0; i < N; ++i) {
		std::cin >> num_list[i];
	}
	backtracking(0, 0, 0);
	std::cout << max_value << std::endl;
}